cpp/src/lance/format/schema.cc

```diff
@@ -383,6 +385,38 @@ std::shared_ptr<Field> Field::Project(const std::shared_ptr<::arrow::Field>& arr
   return new_field;
 }
 
+::arrow::Result<std::shared_ptr<Field>> Field::Merge(const ::arrow::Field& arrow_field) const {
+  if (name() != arrow_field.name()) {
+    return ::arrow::Status::Invalid(
+        "Attempt to merge two different fields: ", name(), "!=", arrow_field.name());
+  }
+  auto self_type = type();
+  if (self_type->id() != arrow_field.type()->id()) {
+    return ::arrow::Status::Invalid(
+        "Can not merge two fields: ", self_type->ToString(), "!=", arrow_field.type()->ToString());
+  };
+  auto new_field = Copy(true);
+  if (::arrow::is_list_like(self_type->id())) {
+    auto list_type = std::dynamic_pointer_cast<::arrow::ListType>(arrow_field.type());
+    return children_[0]->Merge(*list_type->value_field());
```

Comment:
I don't quite understand this, if we're merging two list fields, then should the return value be a list type of the merge of the two value types?

Reply:
this is the case for merging `list<struct<....>>`, so the `struct` type is nested value type for a list.